File: backend/tests/authenticationTests/registerModuleTests.py
import pytest
import requests

# Base URL of the backend server
BASE_URL = "http://localhost:8000/auth"

# Test data
NEW_USER = {
    "username": "newuser",
    "password": "newpassword"
}

EXISTING_USER = {
    "username": "admin",
    "password": "admin"
}

@pytest.fixture
def setup_existing_user():
    """
    Fixture to set up an existing user in the database.
    This assumes you have an endpoint to create users for testing purposes.
    """
    # Create an existing user
    response = requests.post(f"{BASE_URL}/register", json=EXISTING_USER)
    assert response.status_code == 200  # Ensure the user is created successfully
    yield


def test_register_success():
    """
    Test successful registration of a new user.
    """
    response = requests.post(f"{BASE_URL}/register", json=NEW_USER)
    assert response.status_code == 200
    assert response.json()["data"]["type"] == "Success"
    assert response.json()["data"]["message"] == "User registered successfully"

    # Clean up the newly registered user
    requests.delete(f"{BASE_URL}/delete", json={"username": NEW_USER["username"]})


def test_register_existing_user(setup_existing_user):
    """
    Test registration with an existing username.
    """
    response = requests.post(f"{BASE_URL}/register", json=EXISTING_USER)
    assert response.status_code == 400
    assert response.json()["error"]["type"] == "BadRequest"
    assert response.json()["error"]["message"] == "User already exists"


def test_register_missing_fields():
    """
    Test registration with missing fields in the request body.
    """
    response = requests.post(f"{BASE_URL}/register", json={"username": "newuser"})
    assert response.status_code == 400
    assert response.json()["error"]["type"] == "BadRequest"


def test_register_server_error(monkeypatch):
    """
    Test registration when a server error occurs.
    """
    def mock_save(*args, **kwargs):
        raise Exception("Database error")

    # Mock the save method to simulate a server error
    monkeypatch.setattr("backend.authentication.authenticationModel.User.save", mock_save)

    response = requests.post(f"{BASE_URL}/register", json=NEW_USER)
    assert response.status_code == 500
    assert response.json()["error"]["type"] == "InternalServerError"
    assert response.json()["error"]["message"] == "Server error"

File: backend/tests/authenticationTests/registerModuleTests.js
const register = require('../../backend/authentication/authenticationModules/registerModule');
const User = require('../../backend/authentication/authenticationModel');
const { badRequestJsonResponse, internalErrorJsonResponse, successJsonResponse } = require('../../backend/utils/jsonResponses/jsonResponses');

jest.mock('../../backend/authentication/authenticationModel'); // Mock the User model

describe('Register Module', () => {
  let req, res;

  beforeEach(() => {
    // Mock request and response objects
    req = {
      body: {
        username: 'testuser',
        password: 'testpassword',
      },
    };
    res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn(),
    };
  });

  it('should return 500 if a server error occurs', async () => {
    /**
     * Test registration when a server error occurs.
     */
    // Mock the save method to throw an error
    User.prototype.save = jest.fn().mockImplementation(() => {
      throw new Error('Database error');
    });

    await register(req, res);

    // Assert that the response status is 500
    expect(res.status).toHaveBeenCalledWith(500);

    // Assert that the response contains the correct error message
    expect(res.json).toHaveBeenCalledWith(internalErrorJsonResponse('Server error'));
  });
});
